tests/test_directives.py

```diff
@@ -0,0 +1,270 @@
+    """
+    Tests for _DirectivesItem class
+    """
+    def test_default(self):
+        assert NP.default == 1
+        assert NGPU.default == 0
+        assert NRANKS.default == 0
+        assert OMP_NUM_THREADS.default == 0
+        assert EXECUTABLE.default == sys.executable
+        assert WALLTIME.default == 12.0
+        assert MEMORY.default == 4
+        assert PROCESS_FRACTION.default == 1.
+
+    def test_invalid_values(self, available_directives_list):
+        invalid_values = {'np': 0, 'ngpu': 'ngpu', 'nranks': -1,
+                          'omp_num_threads': -20, 'executable': 1,
+                          'walltime': None, 'memory': 0, 'processor_fraction': 2.5}
+        error_types = {'np': ValueError, 'ngpu': TypeError,
+                       'nranks': ValueError, 'omp_num_threads': ValueError,
+                       'executable': TypeError, 'walltime': TypeError,
+                       'memory': ValueError, 'processor_fraction': ValueError}
+        for dir in available_directives_list:
+            with pytest.raises(error_types[dir.name]):
+                dir.validation(invalid_values[dir.name])
+
+    def test_defaults_are_valid(self, available_directives_list):
+        for directive in available_directives_list:
+            directive.validation(directive.default)
+
+    def test_serial(self):
+        assert NP.serial(4, 2) == 4
+        assert NRANKS.serial(4, 2) == 4
+        assert NGPU.serial(4, 2) == 4
+        assert OMP_NUM_THREADS.serial(4, 2) == 4
+        assert EXECUTABLE.serial('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.serial(4, 2) == 6
+        assert MEMORY.serial(4, 2) == 4
+        assert PROCESS_FRACTION.serial(0.4, 0.2) == 0.4
+
+    def test_parallel(self):
+        assert NP.parallel(4, 2) == 6
+        assert NRANKS.parallel(4, 2) == 6
+        assert NGPU.parallel(4, 2) == 6
+        assert OMP_NUM_THREADS.parallel(4, 2) == 6
+        assert EXECUTABLE.parallel('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.parallel(4, 2) == 4
+        assert MEMORY.parallel(4, 2) == 6
+        assert PROCESS_FRACTION.parallel(0.4, 0.2) == 0.4
+
+    def test_finalize(self):
+        dict_directives = {'nranks': NRANKS.default, 'omp_num_threads': OMP_NUM_THREADS.default}
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 8
+        dict_directives['nranks'] = lambda x: x**2
+        assert NP.finalize(2, dict_directives) == 2
+
+    def test_manual_item_default(self, product_directive):
+        assert product_directive.default == 10
+
+    def test_manual_item_validation(self, product_directive):
+        try:
+            val = product_directive.validation(product_directive.default)
+            assert product_directive.default == val
+        except Exception:
+            raise ValueError("Invalid default value for product_directive")
```

Comment:
No need for the try and except blocks here. We are raising regardless. pytest's will show us the line which the exception is raised.
```suggestion
        val = product_directive.validation(product_directive.default)
        assert product_directive.default == val
```

Reply:
Noted